pkg/web/api/health_checker.go

```diff
@@ -17,28 +17,76 @@ package api
+	grpcstatus "google.golang.org/grpc/status"
 )
 
-type HealthChecker struct {
+type Checker interface {
+	Check(ctx context.Context) error
+}
+
+// HealthServer implements the gRPC health service.
+// Using the HealthServer, it's possible to check
+// the gRPC server's overall health, but also the
+// health of the individual gRPC services.
+// For more information, see: https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+type HealthServer struct {
 	grpc_health_v1.UnimplementedHealthServer
+	checkers map[string]Checker
+	log      log.CtxLogger
 }
 
-func (s *HealthChecker) Check(ctx context.Context, req *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
-	// this is intentionally a placeholder for checking Conduit health.
-	// todo: should be changed to check the connection to the storage layer.
-	return &grpc_health_v1.HealthCheckResponse{
-		Status: grpc_health_v1.HealthCheckResponse_SERVING,
-	}, nil
+func (h *HealthServer) Check(ctx context.Context, req *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
+	// Check all services
+	if req.Service == "" {
+		if h.checkAll(ctx) {
+			return &grpc_health_v1.HealthCheckResponse{Status: grpc_health_v1.HealthCheckResponse_SERVING}, nil
+		}
+		return &grpc_health_v1.HealthCheckResponse{Status: grpc_health_v1.HealthCheckResponse_NOT_SERVING}, nil
+	}
+
+	if _, ok := h.checkers[req.Service]; ok {
+		if h.check(ctx, req.Service) {
+			return &grpc_health_v1.HealthCheckResponse{Status: grpc_health_v1.HealthCheckResponse_SERVING}, nil
+		}
+		return &grpc_health_v1.HealthCheckResponse{Status: grpc_health_v1.HealthCheckResponse_NOT_SERVING}, nil
+	}
+
+	return nil, grpcstatus.Errorf(codes.NotFound, "service '%v' not found", req.Service)
 }
 
-func (s *HealthChecker) Watch(req *grpc_health_v1.HealthCheckRequest, server grpc_health_v1.Health_WatchServer) error {
+func (h *HealthServer) Watch(req *grpc_health_v1.HealthCheckRequest, server grpc_health_v1.Health_WatchServer) error {
 	// should be altered to subsequently send a new message whenever the service's serving status changes.
 	return server.Send(&grpc_health_v1.HealthCheckResponse{
 		Status: grpc_health_v1.HealthCheckResponse_SERVING,
 	})
 }
 
-func NewHealthChecker() *HealthChecker {
-	return &HealthChecker{}
+// checkAll checks returns `true` if all service are healthy,
+// returns `false` otherwise (one or more services are not healthy).
+func (h *HealthServer) checkAll(ctx context.Context) bool {
+	ok := true
+	for service := range h.checkers {
+		ok = ok && h.check(ctx, service)
+	}
+	return ok
+}
+
+func (h *HealthServer) check(ctx context.Context, service string) bool {
+	err := h.checkers[service].Check(ctx)
+	if err != nil {
+		h.log.Err(ctx, err).
+			Str("service_name", service).
+			Msg("health check not OK")
+	}
+	return err == nil
+}
+
+func NewHealthServer(checkers map[string]Checker, log log.CtxLogger) *HealthServer {
```

Comment:
Nitpick: normally I'd expect the constructor to be placed directly after the type declaration, I think [this is a good guideline](https://github.com/uber-go/guide/blob/master/style.md#function-grouping-and-ordering).

Reply:
Definitely agree. I normally do it but as new methods were added through the IDE they kept appearing at the top.:)